### PR TITLE
Add renderProp to ShallowWrapper

### LINKED ISSUE #1444: Feature request: wrap JSX without any rendering
Sometimes I find that I need to run assertions on raw `ReactElement`s, without needing to shallow render or mount. This is especially common with the hip new "render prop" pattern. Here's an example based on React Router:

```js
// Using Route's render prop to inject the `match`, as seen in this example:
// https://reacttraining.com/react-router/web/example/custom-link
const MyCustomLink = (children, to) => (
  <Route
    render={({ match }) => (
      <ObnoxiousHighlight on={Boolean(match)}>
        <Link to={to}>{children}</Link>
      </ObnoxiousHighlight>
    )}
  />
);

describe('MyCustomLink', () => {
  it('renders', () => {
    const outerWrapper = shallow(<MyCustomLink to="/home">Home</MyCustomLink>);
    const route = outerWrapper.find(Route);
    const innerJsx = route.prop('render')({ match: { __testMatch: true } });
    const innerWrapper = __someNewEnzymeFunction(innerJsx);
    expect(innerWrapper.find(ObnoxiousHighlight)).toHaveProp('on', true);
    expect(innerWrapper.find(Link)).toHaveProp('to', '/home');
  });
});
```

Note that if I called `shallow(innerJsx)`, it would be treated as a shallow render of `ObnoxiousHighlight`, which isn't what I'm interested in testing in this suite.

As a workaround, I've been using this dirty function:
```js
const wrap = element => shallow(<div>{element}</div>).childAt(0);
```

But it would be nice to have it as part of Enzyme.

### PR DESCRIPTION
**Pull request by dferber90**
Adds a `renderProp` method to `ShallowWrapper`s to ease testing render props.

This PR basically extracts `renderProp`  out of [@commercetools/enzyme-extensions](https://github.com/commercetools/enzyme-extensions). We would deprecate and drop it there if `renderProp` makes it into `enzyme` itself.

I wrote [this article](https://medium.com/@dferber90/test-a-render-prop-6a44e02f4c39) a while ago detailing how to use `renderProp` and why.

I'm curious for feedback of whether you'd want to accept it or not after some interest was shown in https://github.com/commercetools/enzyme-extensions/issues/8.

---

Thanks to @tdeekens who helped set up and bounce ideas back&forth with for `renderProp` in `enzyme-extensions`!

Closes #1444. Related to #1856.

### CODE REVIEW

**Comment by dferber90** [comment]
The tests are fine for the most part. However, I keep getting errors in React 0.13 and 0.14, and I'm not sure how to resolve them. Maybe somebody has seen them before and knows what to do?


React 0.14

```
  1) shallow .renderProp() returns a ShallowWrapper wrapper around the node returned from the render prop:
     TypeError: inst.render is not a function
      at ShallowComponentWrapper._renderValidatedComponentWithoutOwnerOrContext (node_modules/react/lib/ReactCompositeComponent.js:587:34)
      at ShallowComponentWrapper.mountComponent (node_modules/react/lib/ReactCompositeComponent.js:220:30)
      at ShallowComponentWrapper.wrapper [as mountComponent] (node_modules/react/lib/ReactPerf.js:66:21)
      at ReactShallowRenderer._render (node_modules/react/lib/ReactTestUtils.js:366:14)
      at _batchedRender (node_modules/react/lib/ReactTestUtils.js:348:12)
      at ReactDefaultBatchingStrategyTransaction.perform (node_modules/react/lib/Transaction.js:136:20)
      at Object.batchedUpdates (node_modules/react/lib/ReactDefaultBatchingStrategy.js:62:19)
      at Object.batchedUpdates (node_modules/react/lib/ReactUpdates.js:94:20)
      at ReactShallowRenderer.render (node_modules/react/lib/ReactTestUtils.js:343:16)
      at packages/enzyme-adapter-react-14/build/ReactFourteenAdapter.js:1:1
      at withSetStateAllowed (packages/enzyme-adapter-utils/build/Utils.js:1:1)
      at Object.render (packages/enzyme-adapter-react-14/build/ReactFourteenAdapter.js:1:1)
      at new ShallowWrapper (packages/enzyme/build/ShallowWrapper.js:1:1)
      at ShallowWrapper.wrap (packages/enzyme/build/ShallowWrapper.js:1:1)
      at ShallowWrapper.<anonymous> (packages/enzyme/build/ShallowWrapper.js:1:1)
      at ShallowWrapper.single (packages/enzyme/build/ShallowWrapper.js:1:1)
      at ShallowWrapper.shallow (packages/enzyme/build/ShallowWrapper.js:1:1)
      at ShallowWrapper.<anonymous> (packages/enzyme/build/ShallowWrapper.js:1:1)
      at ShallowWrapper.single (packages/enzyme/build/ShallowWrapper.js:1:1)
      at ShallowWrapper.renderProp (packages/enzyme/build/ShallowWrapper.js:1:1)
      at Context.<anonymous> (packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx:4759:26)
      at process.topLevelDomainCallback (domain.js:121:23)
```

React 0.13

```
1) shallow .renderProp() returns a ShallowWrapper wrapper around the node returned from the render prop:
     TypeError: Cannot add property context, object is not extensible
      at ShallowComponentWrapper.mountComponent (node_modules/react/lib/ReactCompositeComponent.js:153:18)
      at ShallowComponentWrapper.wrapper [as mountComponent] (node_modules/react/lib/ReactPerf.js:70:21)
      at ReactShallowRenderer._render (node_modules/react/lib/ReactTestUtils.js:391:14)
      at ReactShallowRenderer.render (node_modules/react/lib/ReactTestUtils.js:375:8)
      at ReactShallowRenderer.contextCompatibleRender [as render] (packages/enzyme-adapter-react-13/build/ReactThirteenAdapter.js:1:1)
      at Object.render (packages/enzyme-adapter-react-13/build/ReactThirteenAdapter.js:1:1)
      at new ShallowWrapper (packages/enzyme/build/ShallowWrapper.js:1:1)
      at ShallowWrapper.wrap (packages/enzyme/build/ShallowWrapper.js:1:1)
      at ShallowWrapper.<anonymous> (packages/enzyme/build/ShallowWrapper.js:1:1)
      at ShallowWrapper.single (packages/enzyme/build/ShallowWrapper.js:1:1)
      at ShallowWrapper.shallow (packages/enzyme/build/ShallowWrapper.js:1:1)
      at ShallowWrapper.<anonymous> (packages/enzyme/build/ShallowWrapper.js:1:1)
      at ShallowWrapper.single (packages/enzyme/build/ShallowWrapper.js:1:1)
      at ShallowWrapper.renderProp (packages/enzyme/build/ShallowWrapper.js:1:1)
      at Context.<anonymous> (packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx:4752:33)
      at process.topLevelDomainCallback (domain.js:121:23)
```

File: packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx [hunk 1]
```diff
@@ -4743,6 +4743,24 @@ describe('shallow', () => {
     });
   });
 
+  describe('.renderProp()', () => {
+    it('returns a ShallowWrapper wrapper around the node returned from the render prop', () => {
+      const Foo = () => (<div className="in-foo" />);
+      const Bar = props => (<div className="in-bar">{props.render()}</div>);
```
Comment (ljharb) [inline packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx]: let's use class-based components here, so that it'll pass in react 13.

File: packages/enzyme/src/ShallowWrapper.js [hunk 2]
```diff
@@ -220,7 +220,7 @@ class ShallowWrapper {
           });
         }
       }
-    // creating a child component through enzyme's ShallowWrapper APIs.
+      // creating a child component through enzyme's ShallowWrapper APIs.
```
Comment (ljharb) [inline packages/enzyme/src/ShallowWrapper.js]: let's revert these unrelated indentation changes

File: packages/enzyme/src/ShallowWrapper.js [hunk 3]
```diff
@@ -1033,6 +1033,34 @@ class ShallowWrapper {
...
   }
 
+  /**
+   * Returns a wrapper of the node rendered by the provided render prop.
+   *
+   * @param {String} propName
+   * @returns {ShallowWrapper}
+   */
+  renderProp(propName, ...args) {
+    const Wrapper = props => props.children;
```
Comment (ljharb) [inline packages/enzyme/src/ShallowWrapper.js]: This won't work in React 13.

File: packages/enzyme/src/ShallowWrapper.js [hunk 4]
```diff
@@ -1033,6 +1033,34 @@ class ShallowWrapper {
...
+    return this.single('renderProp', () => {
+      if (typeof propName !== 'string') {
+        throw new TypeError('`propName` must be a string');
+      }
+      const prop = this.prop(propName);
+      if (!prop) {
```
Comment (ljharb) [inline packages/enzyme/src/ShallowWrapper.js]: `prop` could be falsy with `0`, `false`, and a number of other options. Perhaps instead this should do `const props = this.props()`, and then do `has(props, propName)`?

File: packages/enzyme/src/ShallowWrapper.js [hunk 5]
```diff
@@ -1033,6 +1033,34 @@ class ShallowWrapper {
...
+        throw new Error(`no prop called “${propName}“ found`);
+      }
+      if (typeof prop !== 'function') {
+        throw new Error(
```
Comment (ljharb) [inline packages/enzyme/src/ShallowWrapper.js]: this should be a TypeError

File: packages/enzyme/src/ShallowWrapper.js [hunk 6]
```diff
@@ -1033,6 +1033,34 @@ class ShallowWrapper {
...
+          `expected prop “${propName}“ to contain a function, but it holds “${typeof prop}“`,
+        );
+      }
+
+      const element = prop(...args);
+      const wrapped = getAdapter(this[OPTIONS]).createElement(Wrapper, null, element);
```
Comment (ljharb) [inline packages/enzyme/src/ShallowWrapper.js]: we may need to add a method to all the adapters "wrap", so that the adapter itself can handle the wrapper implementation.

File: packages/enzyme/src/ShallowWrapper.js (hunk 4 above)
Comment (ljharb) [reply packages/enzyme/src/ShallowWrapper.js]: that's a reasonable alternative; altho it's nice to be able to give a better error message "this prop doesn't exist" versus "this prop isn't a function".

File: packages/enzyme/src/ShallowWrapper.js (hunk 6 above)
Comment (ljharb) [reply packages/enzyme/src/ShallowWrapper.js]: I think it'd need to be in this PR (in a separate commit, ofc).

**Comment by dferber90** [comment]
> I tried to add `wrap` to all adapters but I'm not familiar enough with the internals of `enzyme` to get the tests to pass. I still keep bumping into `TypeError: inst.render is not a function` on React 14 😞  Other versions of React seem to be passing the tests now though!

The issue was `const stub = sinon.stub().returns(null);` effectively rendering `null` from a stateless functional component, which is not supported in React 14. Changing it to `const stub = sinon.stub().returns(<div />);` solved the issue. [This comment](https://github.com/styleguidist/react-styleguidist/issues/498#issue-236608981) tipped me off.

File: packages/enzyme-adapter-react-13/src/ReactThirteenAdapter.js [hunk 7]
```diff
@@ -242,6 +242,17 @@ class ReactThirteenAdapter extends EnzymeAdapter {
...
+  wrap(element) {
+    class Wrapper extends React.Component {
+      render() {
+        // eslint-disable-next-line react/prop-types
+        const { children } = this.props;
+        return children;
+      }
+    }
+    return React.createElement(Wrapper, null, element);
+  }
```
Comment (dferber90) [inline packages/enzyme-adapter-react-13/src/ReactThirteenAdapter.js]: @ljharb Could you confirm this is what you had in mind for the `wrap` function on the adapters?

**Comment by ljharb** [comment]
I've updated and rebased the branch; I'll give it another review and then we can add the method for `mount` as well.

File: docs/api/ShallowWrapper/renderProp.md [hunk 8]
```diff
@@ -0,0 +1,85 @@
+# `.renderProp(propName, ...args) => ShallowWrapper`
+
+Calls the current wrapper's property with name `propName` and the `args` provided.
+Returns the result in a new shallow wrapper.
+
+NOTE: can only be called on wrapper of a single non-DOM component element node.
```
Comment (ljharb) [inline docs/api/ShallowWrapper/renderProp.md]: where is "non-DOM" being enforced? we might need to check the node type isn't "host" in `renderProp`.

Comment (dferber90) [reply docs/api/ShallowWrapper/renderProp.md]: Nice catch. It wasn't actually enforced so far. I now added an explicit check and amended the last commit. Copied that one from `simulateError`.

**Comment by dferber90** [comment]
I'm still a bit puzzled by how `renderProp` would work with `mount` - or what it would be useful for (I haven't actually used `mount` myself). Depending on the system under test, the `renderProp` will likely get called anyways during rendering, so users can test the final output.

The only useful case for having `renderProp` in `mount` I could come up with would be to influence the arguments the actual render-prop gets called with. But there should be other means to do so by setting the parent component up properly. I'm questioning how useful having `renderProp` in `mount` would be after all.

I attempted to add `mount` anyways, but I failed at coming up with a good example for docs/tests and at actually making the implementation work. I'm not sure how to proceed as I feel kinda lost without understanding the end-goal of having `renderProp` in `mount`.

It would be nice if you could give me some direction or take over from here.

**Comment by ljharb** [comment]
It's so you can unit-test the render prop without needing to know all the permutations of props that lead to certain input combinations.

**Comment by dferber90** [comment]
@ljharb Thanks for driving this home!

I visited a friend this weekend so I didn't get to continue working on it. I really appreciate your help!

We'll deprecate the `renderProp` function in [@commercetools/enzyme-extensions](https://github.com/commercetools/enzyme-extensions) once this is released.

**Comment by dferber90** [comment]
The current API works like this `wrapper.renderProp(propName, arg1, arg2)`.

This will call the prop `propName` on `wrapper` with `arg1` and `arg2`, like `wrapper[propName](arg1, arg2)`.

This API is nice to work with but prevents us from ever adding something like an options object or any other param to `renderProp` itself, as all arguments after the first one get forwarded to the render prop.

It might be a better idea to use an API like `wrapper.renderProp(propName, [arg1, arg2])` or `wrapper.renderProp(propName)(arg1, arg2)`.

So far this has not been a problem for us but I wanted to give a heads-up now that it's becoming official. I'm undecided of which API is suited best.

**Comment by ljharb** [comment]
@dferber90 those are really good points. The explicit array is cleaner, until you want no arguments and you want to pass options. The “return a function” approach is cleaner when you have arguments, but perhaps less so when you don’t.

I think I’m slightly leaning towards the “returning a function” API. I’ll be releasing the adapters sooner than enzyme itself, so we have a bit of time to make the change, but not much :-)

File: docs/api/ReactWrapper/renderProp.md [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+
+expect(wrapper.equals(<h1>The mouse position is 0, 0</h1>)).to.equal(true);
+```
+
+##### Testing with multiple arguments
+
+```jsx
+const wrapper = mount(<App />)
+  .find(Mouse)
+  .renderProp('render', [10, 20]);
```
Comment (dferber90) [inline docs/api/ReactWrapper/renderProp.md]: This is under "multiple arguments", so the code should be `.renderProp('render', 10, 20)`

File: docs/api/ShallowWrapper/renderProp.md [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+expect(wrapper.equals(<h1>The mouse position is 0, 0</h1>)).to.equal(true);
+```
+
+##### Testing with multiple arguments
+
+```jsx
+const wrapper = shallow(<App />)
+  .find(Mouse)
+  .renderProp('render', [10, 20]);
```
Comment (dferber90) [inline docs/api/ShallowWrapper/renderProp.md]: Same here, this should also be `.renderProp('render', 10, 20)`.

File: packages/enzyme-adapter-utils/src/wrapWithSimpleWrapper.jsx [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+        return children;
+      }
+    }
+    SimpleClassWrapper.propTypes = propTypes;
+    return SimpleClassWrapper;
+  }
+)();
+
+export default function wrap(element) {
+  console.log(React.version, Wrapper);
```
Comment (dferber90) [inline packages/enzyme-adapter-utils/src/wrapWithSimpleWrapper.jsx]: Is this a debugging-leftover? This should go away as far as I can tell.

**Comment by dferber90** [comment]
@ljharb I now also opened https://github.com/airbnb/enzyme/pull/1891 to suggest the "return a function" approach instead.
